chainer/training/updater.py

```diff
@@ -0,0 +1,366 @@
+    def __init__(self, iterator, optimizer, update_func=None, converter=None,
+                 models=None, devices=None):
+        super(ParallelUpdater, self).__init__(
+            iterator=iterator,
+            optimizer=optimizer,
+            converter=converter
+        )
+        assert models is not None or devices is not None, \
+            "Either 'models' or 'devices' must be defined."
+
+        if models is None:
+            names = list(six.iterkeys(devices))
+            try:
+                names.remove('main')
+            except ValueError:
+                raise KeyError("'devices' must contain a 'main' key.")
+            models = {'main': optimizer.target}
+            for name in names:
+                models[name] = models['main'].copy()
+
+        if devices is None:
+            devices = dict((name, i) for i, name in enumerate(names))
+
+        for name, model in models.items():
+            model.to_gpu(devices[name])
+
+        self._models = models
+
+        self._update_func = update_func or _parallel_update(
+            self, converter, models, devices)
+
+    def connect_trainer(self, trainer):
+        """Connects the updater to the trainer that will call it.
+
+        The typical usage of this method is to register additional links to the
+        reporter of the trainer. This method is called at the end of the
+        initialization of :class:`~chainer.training.Trainer`. The default
+        implementation does nothing.
+
+        Args:
+            trainer (~chainer.training.Trainer): Trainer object to which the
+                updater is registered.
+
+        """
+        # Add observers for all (other) models.
+        model_main = self.get_optimizer('main').target
+        models_others = {
+            k: v for k, v in self._models.items() if v != model_main
+        }
+        for name, model in models_others.items():
+            trainer.reporter.add_observer(name, model)
+
+
+def _parallel_update(updater, converter, models, devices):
+
+    if not converter:
+        converter = convert.concat_examples
+
+    iterator = updater.get_iterator('main')
+    optimizer = updater.get_optimizer('main')
+    model_main = optimizer.target
+    models_others = {k: v for k, v in models.items() if v != model_main}
+
+    def update(_):
+        batch = iterator.next()
+
+        #
+        # Split the batch to sub-batches.
+        #
+        n = len(models.items())
+        in_arrays_list = {}
+        for i, key in enumerate(models.keys()):
+            in_arrays_list[key] = converter(batch[i::n], devices[key])
+
+        for model in models.values():
+            model.zerograds()
+
+        losses = []
+        for model_key in models.keys():
+            model = models[model_key]
+            in_arrays = in_arrays_list[model_key]
+
+            if isinstance(in_arrays, tuple):
+                in_vars = tuple(variable.Variable(x) for x in in_arrays)
+                losses.append(model(*in_vars))
```

Comment:
If the GPU computation of forward is heavy enough, yes. It would be better to use threading to better parallelize the GPU computation.